lib/src/editor.dart

```diff
@@ -0,0 +1,790 @@
+import 'package:yaml/src/equality.dart';
+import 'package:yaml/yaml.dart';
+
+import './source_edit.dart';
+
+/// An interface for modififying [YAML][1] documents while preserving comments and
+/// whitespaces.
+///
+/// YAML parsing is supported by `package:yaml`, and modifications are performed as
+/// string operations. Each time a modification takes place via one of the public
+/// methods, we calculate the expected final result, and parse the result YAML string,
+/// and ensure the two YAML trees match. Users may define the default settings to be
+/// applied to these string modifications. Note however that these settings only apply
+/// to portions of the YAML that are modified by this class.
+///
+/// Most modification methods require the user to pass in an [Iterable<Object>] path that
+/// holds the keys/indices to navigate to the element. Key equality is performed via
+/// `package:yaml`'s [YamlMap]'s key equality.
+///
+/// [1]: https://yaml.org/
+class YamlEditor {
+  /// List of [SourceEdit]s that have been applied to [_yaml] since the creation of this
+  /// instance, in chronological order. Intended to be compatible with `package:analysis_server`.
+  final List<SourceEdit> edits = [];
```

Comment:
We could consider just having top-level functions returning a single SourceEdit:
```
SouceEdit setIn(String yaml, Iterable<Object> path, Object value);
```

It is not clear how much we gain by accumulating the edits as long as we re-parse the yaml for every step.

Reply:
Well, we manage to hide the re-parse logic..